### PR TITLE
Distinguish between North and South detectors by AGET ID

### PR DESCRIPTION
**Pull request by DavidDiezIb**
DavidDiezIb Large: 253 [](https://gitlab.cern.ch/rest-for-physics/rawlib/-/commits/david_TREXsides) [](https://gitlab.cern.ch/rest-for-physics/framework/-/commits/david_TREXsides)

New process added to distinguish events from North and South sides using only the AGET ID.  Not readout needed.

### CODE REVIEW

**Comment by jgalan** [comment]
That makes the process less specific, and you define how to use it at the RML.

**Comment by DavidDiezIb** [comment]
The purpose of this is precisely to avoid the use of the readout, as a way of testing other processes that make use of itto extract the same information (could be done with `zMean` observable from `TRestDetectorHitsAnalysisProcess`).

I agree it could be generalized with the ID ranges, I'll change it. At first thought, I prefered to set just one number but Now I see it is easier to read the code (in the rml) if we use the ranges.

**Comment by nkx111** [comment]
Why no East and West :P? 

The process name also looks specific. We shall not include experiment names in it.

**Comment by jgalan** [comment]
> Why no East and West :P?
> 
> The process name also looks specific. We shall not include experiment names in it.

Right, the day that TREX-DM is moved to a new hall and placed in west-east direction you will need to change the parameter names.

**Comment by DavidDiezIb** [comment]
You are right, I'll remove all references to TREX experiment. And I'll try to implement as you suggested @jgalan 


> ```
> <TRestRawDaqIdTaggingProcess ...>
>       <tag name="South" ids="(0,72*8-1)" />
>       <tag name="North" ids="(72*8,72*16)" />
> </TRestRawDaqIdTaggingProcess>
> ```

But how should I define these "tag" sections? I don't know how write that in order to the rml understands the parameters, do we have any process that use something similar?

**Comment by jgalan** [comment]
You may have a look to the following example: https://github.com/rest-for-physics/geant4lib/blob/master/src/TRestGeant4Metadata.cxx where it is extracted `activeVolume`,  having a look it should be something as follows in the implementation of `InitFromConfigFile`.

There, it is extracted the `qualityNumber` xml keyword. You do a for loop to extract every line found. It would be something similar to this:

```
void ... InitFromConfigFile() ...


     // This line is to exploit the retrieval of parameter as it is done in each process
     TRestEventProcess::InitFromConfigFile();

    // This is the additional code required by the process to read tags
    TiXmlElement* tagDefinition = GetElement("tag");
    while (tagDefinition) {
        fTagNames.push_back( GetFieldValue("name", tagDefinition)) );

        fIdRanges.push_back( Get2DVectorParameterWithUnits( "ids", tagDefinition );

        }
        tagDefinition = GetNextElement(tagDefinition);
    }
```

**Comment by jgalan** [comment]
I find the above `tagCode` the best way, but you need then to look into the metadata and understand what it means each code. It takes a bit of additional effort, but then you may see the number of events that gave something only in South (tagCode=1), the number of events that gave something in North (tagCode=2), and the number of events that gave something in Both (tagCode=3).

**Comment by DavidDiezIb** [comment]
Thaks for the help @jgalan. Now the process should be completely generic, it admits any number of daq Id ranges that can be named as we wish. Basically I implemented your proposed strategy and it worked really well.

File: src/TRestRawSignalDaqIdTaggingProcess.cxx [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    // This is the additional code required by the process to read tags
+    TiXmlElement* tagDefinition = GetElement("tag");
+    while (tagDefinition) {
+        fTagNames.push_back( GetFieldValue("name", tagDefinition) );
+        fIdRanges.push_back( Get2DVectorParameterWithUnits("ids", tagDefinition) );
+
+        tagDefinition = GetNextElement(tagDefinition);
+    }
+        
+    PrintMetadata();
```
Comment (jgalan) [inline src/TRestRawSignalDaqIdTaggingProcess.cxx:119]:
```suggestion
    PrintMetadata();

```

Comment (jgalan) [reply src/TRestRawSignalDaqIdTaggingProcess.cxx:119]: `PrintMetadata` is invoked externally during the processing chain when the process verboseLevel is set to `info`. Then, this `PrintMetadata` call will make that the process metadata is printed out several times. When instantiating it by other means, the user then needs to call `PrintMetadata` manually.

**Review by jgalan — APPROVED** [review]

**Pull request merged**